core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,1799 @@
+      this
+    } else {
+      val newElements = {
+        val nonWitUtxoVec =
+          this.getRecords[NonWitnessOrUnknownUTXO](NonWitnessUTXOKeyId)
+        val witScriptVec =
+          this.getRecords[InputPSBTRecord.WitnessScript](WitnessScriptKeyId)
+        if (nonWitUtxoVec.size == 1 && witScriptVec.size == 1) {
+          val nonWitUtxo = nonWitUtxoVec.head.transactionSpent
+          if (txIn.previousOutput.vout.toInt < nonWitUtxo.outputs.size) {
+            val out = nonWitUtxo.outputs(txIn.previousOutput.vout.toInt)
+            elements
+              .filterNot(_.key.head == NonWitnessUTXOKeyId.byte)
+              .filterNot(_.key.head == WitnessUTXOKeyId.byte) :+ WitnessUTXO(
+              out)
+          } else {
+            elements.filterNot(_.key.head == NonWitnessUTXOKeyId.byte)
+          }
+        } else {
+          elements
+        }
+      }
+      InputPSBTMap(newElements)
+    }
+  }
+}
+
+object InputPSBTMap {
+
+  /** Constructs a finalized InputPSBTMap from a UTXOSpendingInfoFull,
+    * the corresponding PSBT's unsigned transaction, and if this is
+    * a non-witness spend, the transaction being spent
+    */
+  def finalizedFromUTXOSpendingInfo(
+      spendingInfo: UTXOSpendingInfoFull,
+      unsignedTx: Transaction,
+      nonWitnessTxOpt: Option[Transaction])(
+      implicit ec: ExecutionContext): Future[InputPSBTMap] = {
+    val sigComponentF = BitcoinSigner
+      .sign(spendingInfo, unsignedTx, isDummySignature = false)
+
+    sigComponentF.map { sigComponent =>
+      val utxos = spendingInfo match {
+        case _: SegwitV0NativeUTXOSpendingInfoFull |
+            _: P2SHNestedSegwitV0UTXOSpendingInfoFull =>
+          Vector(WitnessUTXO(spendingInfo.output))
+        case _: RawScriptUTXOSpendingInfoFull | _: P2SHNoNestSpendingInfo |
+            _: UnassignedSegwitNativeUTXOSpendingInfo =>
+          nonWitnessTxOpt match {
+            case None => Vector.empty
+            case Some(nonWitnessTx) =>
+              Vector(NonWitnessOrUnknownUTXO(nonWitnessTx))
+          }
+      }
+
+      val scriptSig =
+        FinalizedScriptSig(sigComponent.scriptSignature)
+      sigComponent.transaction match {
+        case _: BaseTransaction => InputPSBTMap(utxos ++ Vector(scriptSig))
+        case wtx: WitnessTransaction =>
+          val witness = wtx.witness.witnesses(sigComponent.inputIndex.toInt)
+          val scriptWitness = FinalizedScriptWitness(witness)
+          val finalizedSigs =
+            if (witness != EmptyScriptWitness) {
+              Vector(scriptSig, scriptWitness)
+            } else {
+              Vector(scriptSig)
+            }
+          InputPSBTMap(utxos ++ finalizedSigs)
+      }
+    }
+  }
+
+  /** Constructs a full (ready to be finalized) but unfinalized InputPSBTMap
+    * from a UTXOSpendingInfoFull, the corresponding PSBT's unsigned transaction,
+    * and if this is a non-witness spend, the transaction being spent
+    */
+  def fromUTXOSpendingInfo(
```

Comment:
this entire function should be re-written to use `spendingInfo.toSingles` rather than injecting side-effects into `Sign.signFunction`